cliquet/storage/postgresql/__init__.py

```diff
@@ -0,0 +1,458 @@
+            result = cursor.fetchone()
+        encoding = result['encoding'].lower()
+        assert encoding == 'utf8', 'Unexpected database encoding %s' % encoding
+
+        # Create schema
+        here = os.path.abspath(os.path.dirname(__file__))
+        schema = open(os.path.join(here, 'schema.sql')).read()
+        with self.connect() as cursor:
+            cursor.execute(schema)
+        logger.info('Created PostgreSQL storage tables')
+
+    def flush(self):
+        """Delete records from tables without destroying schema. Mainly used
+        in tests suites.
+        """
+        query = """
+        DELETE FROM deleted;
+        DELETE FROM records;
+        """
+        with self.connect() as cursor:
+            cursor.execute(query)
+        logger.debug('Flushed PostgreSQL storage tables')
+
+    def ping(self):
+        try:
+            with self.connect() as cursor:
+                cursor.execute("SELECT now();")
+            return True
+        except psycopg2.Error:
+            return False
+
+    def collection_timestamp(self, resource, user_id):
+        query = """
+        SELECT resource_timestamp(%(user_id)s, %(resource_name)s)::BIGINT
+            AS timestamp;
+        """
+        resource_name = classname(resource)
+        placeholders = dict(user_id=user_id, resource_name=resource_name)
+        with self.connect() as cursor:
+            cursor.execute(query, placeholders)
+            result = cursor.fetchone()
+        return result['timestamp']
+
+    def create(self, resource, user_id, record):
+        query = """
+        INSERT INTO records (user_id, resource_name, data)
+        VALUES (%(user_id)s, %(resource_name)s, %(data)s::json)
+        RETURNING id, last_modified::BIGINT
+        """
+        resource_name = classname(resource)
+        placeholders = dict(user_id=user_id,
+                            resource_name=resource_name,
+                            data=json.dumps(record))
+
+        with self.connect() as cursor:
+            self.check_unicity(resource, user_id, record)
+
+            cursor.execute(query, placeholders)
+            inserted = cursor.fetchone()
+
+        record = record.copy()
+        record[resource.id_field] = inserted['id']
+        record[resource.modified_field] = inserted['last_modified']
+        return record
+
+    def get(self, resource, user_id, record_id):
+        query = """
+        SELECT last_modified::BIGINT, data
+          FROM records
+         WHERE id = %(record_id)s
+        """
+        placeholders = dict(record_id=record_id)
+        with self.connect() as cursor:
+            cursor.execute(query, placeholders)
+            if cursor.rowcount == 0:
+                raise exceptions.RecordNotFoundError(record_id)
+            else:
+                result = cursor.fetchone()
+
+        record = result['data']
+        record[resource.id_field] = record_id
+        record[resource.modified_field] = result['last_modified']
```

Comment:
I wonder if we should factorize this out of this database, since we're duplicating (I think) these assignments across our storage implementations